espnet2/enh/layers/dnn_wpe.py

```diff
@@ -84,11 +84,31 @@ def forward(
                 delay=self.delay,
                 inverse_power=self.inverse_power,
             )
-            enhanced = enhanced.type(data.dtype)
+            enhanced = enhanced.to(dtype=data.dtype)
             enhanced.masked_fill_(make_pad_mask(ilens, enhanced.real), 0)
 
         # (B, F, C, T) -> (B, T, C, F)
         enhanced = enhanced.permute(0, 3, 2, 1)
         if mask is not None:
             mask = mask.transpose(-1, -3)
         return enhanced, ilens, mask
+
+    def predict_mask(
+        self, data: ComplexTensor, ilens: torch.LongTensor
+    ) -> Tuple[torch.Tensor, torch.LongTensor]:
+        """Predict mask for WPE dereverberation
+
+        Args:
+            data (ComplexTensor): (B, T, C, F), double precision
+            ilens (torch.Tensor): (B,)
+        Returns:
+            masks (torch.Tensor): (B, T, C, F)
+            ilens (torch.Tensor): (B,)
+        """
+        if self.use_dnn_mask:
+            (mask,), ilens = self.mask_est(data.permute(0, 3, 2, 1).float(), ilens)
+            # (B, F, C, T) -> (B, T, C, F)
+            mask = mask.transpose(-1, -3)
+        else:
+            mask = None
+        return mask, ilens
```

Comment:
In this file, you seem to be using mask_estimator from espnet1. I think using sigmoid activation might not be best for dereverberation mask. I think torch.clamp would be better. Can we pass arguments to control this activation instead of sigmoid being the only option?

Reply:
Also, for dereverberation, doing a direct signal approximation (instead of masking approximation) is better when trained as a pipeline. Is it possible to add this option?

Reply:
Thank you for your suggestions!

> Can we pass arguments to control this activation instead of sigmoid being the only option?

Yes. I have added two additional arguments to allow choosing the nonlinear type for both wpe and beamformer.

> Also, for dereverberation, doing a direct signal approximation (instead of masking approximation) is better when trained as a pipeline. Is it possible to add this option?

This is a bit complicated. Because the training loss is calculated in `espnet2/enh/espnet_model.py`. For now, it only gives an option to choose the loss type for all masks. If we want to choose the loss type for a specific module of a specific model, the code will be a bit messy.

A current workaround may be training the WPE and beamformer part (using different loss types) separately and combine them later.